Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_inject.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_inject.c	(revision 4293)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_inject.c	(working copy)
@@ -255,7 +255,7 @@
   SIDELOAD_KEXT *CurrentPlugInKext = NULL;
   SIDELOAD_KEXT *Next = NULL;
   
-	if (Entry == 0)/* || OSFLAG_ISUNSET(Entry->Flags, OSFLAG_WITHKEXTS) */ {
+	if ((Entry == 0) || OSFLAG_ISUNSET(Entry->Flags, OSFLAG_WITHKEXTS)) {
 		return EFI_NOT_STARTED;
 	}
   
