Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_patcher.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_patcher.c	(revision 4246)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/kext_patcher.c	(working copy)
@@ -1225,7 +1225,7 @@
   //Slice
   // I see no reason to disable kext injection if FakeSMC found in cache
   //since rev4240 we have manual kext inject disable
- // CheckForFakeSMC(WholePlist, Entry);
+  CheckForFakeSMC(WholePlist, Entry);
   
   DictPtr = WholePlist;
   while ((DictPtr = AsciiStrStr(DictPtr, "dict>")) != NULL) {
@@ -1384,4 +1384,3 @@
     
   }
 }
-
