Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c	(revision 4239)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c	(working copy)
@@ -319,6 +319,7 @@
   DBG("\tDebug: %c\n", Patches->KPDebug ? 'y' : 'n');
   DBG("\tKernelCpu: %c\n", Patches->KPKernelCpu ? 'y' : 'n');
   DBG("\tLapic: %c\n", Patches->KPLapicPanic ? 'y' : 'n');
+  DBG("\tHaswell-E: %c\n", Patches->KPHaswellE ? 'y' : 'n');
   DBG("\tIvy Bridge XCPM: %c\n", Patches->KPIvyXCPM ? 'y' : 'n');
   DBG("\tAICPUPM: %c\n", Patches->KPAsusAICPUPM ? 'y' : 'n');
   DBG("\tAppleRTC: %c\n", Patches->KPAppleRTC ? 'y' : 'n');
