Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c	(revision 4278)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/main.c	(working copy)
@@ -319,6 +319,7 @@
   DBG("\tDebug: %c\n", Patches->KPDebug ? 'y' : 'n');
   DBG("\tKernelCpu: %c\n", Patches->KPKernelCpu ? 'y' : 'n');
   DBG("\tKernelLapic: %c\n", Patches->KPKernelLapic ? 'y' : 'n');
+  DBG("\tHaswell-E: %c\n", Patches->KPHaswellE ? 'y' : 'n');
   DBG("\tKernelXCPM: %c\n", Patches->KPKernelXCPM ? 'y' : 'n');
   DBG("\tKernelPm: %c\n", Patches->KPKernelPm ? 'y' : 'n');
   DBG("\tAppleIntelCPUPM: %c\n", Patches->KPAppleIntelCPUPM ? 'y' : 'n');
