Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c	(revision 4259)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c	(working copy)
@@ -872,11 +872,11 @@
     Patches->KPIvyXCPM = IsPropertyTrue (Prop);
   }
 
-  /*
+
   Prop = GetProperty(DictPointer, "KernelHaswellE");
   if (Prop != NULL || gBootChanged) {
     Patches->KPHaswellE = IsPropertyTrue(Prop);
-  }*/
+  }
 
   Prop = GetProperty (DictPointer, "ATIConnectorsController");
   if (Prop != NULL) {
@@ -1818,7 +1818,7 @@
 
     // InjectKexts default values
     Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_CHECKFAKESMC);
-  //  Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_WITHKEXTS);
+    Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_WITHKEXTS);
 
     Prop = GetProperty (DictPointer, "InjectKexts");
     if (Prop != NULL) {
