Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c	(revision 4293)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/Platform/Settings.c	(working copy)
@@ -934,6 +934,11 @@
     DBG ("FakeCPUID: %x\n", Patches->FakeCPUID);
   }
 
+  Prop = GetProperty(DictPointer, "KernelHaswellE");
+  if (Prop != NULL || gBootChanged) {
+    Patches->KPHaswellE = IsPropertyTrue(Prop);
+  }
+
   Prop = GetProperty (DictPointer, "ATIConnectorsController");
   if (Prop != NULL) {
     UINTN len = 0, i=0;
@@ -1857,7 +1862,7 @@
 
     // InjectKexts default values
     Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_CHECKFAKESMC);
-  //  Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_WITHKEXTS);
+    Entry->Flags = OSFLAG_UNSET(Entry->Flags, OSFLAG_WITHKEXTS);
 
     Prop = GetProperty (DictPointer, "InjectKexts");
     if (Prop != NULL) {
