Index: /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/lib.h
===================================================================
--- /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/lib.h	(revision 4243)
+++ /Users/asurin/src/edk2/Clover/rEFIt_UEFI/refit/lib.h	(working copy)
@@ -192,9 +192,9 @@
 #define OPT_NVDISABLE       (1 << 5)
 #define OPT_SLIDE           (1 << 6)
 #define OPT_POWERNAPOFF     (1 << 7)
-#define OPT_XCPM            (1 << 8) 
-#define OPT_GNOIDLE         (1 << 9) 
-#define OPT_GNOSLEEP        (1 << 10) 
+#define OPT_XCPM            (1 << 8)
+#define OPT_GNOIDLE         (1 << 9)
+#define OPT_GNOSLEEP        (1 << 10)
 #define OPT_GNOMSI          (1 << 11)
 #define OPT_EHCUSB          (1 << 12)
 #define OPT_KEEPSYMS        (1 << 13)
@@ -247,7 +247,7 @@
   AlignCenter,
   AlignUp,
   AlignDown
-  
+
 } ALIGNMENT;
 
 //mouse types
@@ -270,10 +270,10 @@
   EG_IMAGE *Pointer;
   EG_IMAGE *newImage;
   EG_IMAGE *oldImage;
-  
+
   EG_RECT  newPlace;
   EG_RECT  oldPlace;
-  
+
   UINT64	LastClickTime;  //not EFI_TIME
   EFI_SIMPLE_POINTER_STATE    State;
   MOUSE_EVENT MouseEvent;
@@ -288,7 +288,7 @@
   UNIString,
   RadioSwitch,
   CheckBit,
-  
+
 } ITEM_TYPE;
 
 typedef struct {
@@ -462,7 +462,7 @@
   Scale,
   Crop,
   Tile
-  
+
 } SCALING;
 
 typedef struct {
@@ -559,10 +559,11 @@
   BOOLEAN KPDebug;
   BOOLEAN KPKernelCpu;
   BOOLEAN KPLapicPanic;
+  BOOLEAN KPHaswellE;
   BOOLEAN KPIvyXCPM;
   BOOLEAN KPAsusAICPUPM;
   BOOLEAN KPAppleRTC;
-  BOOLEAN KPKernelPm; 
+  BOOLEAN KPKernelPm;
   BOOLEAN KPDELLSMBIOS;  // Dell SMBIOS patch
 //  UINT8   pad[1];
   UINT32  FakeCPUID;
@@ -571,12 +572,12 @@
 #if defined(MDE_CPU_IA32)
   UINT32  align1;
 #endif
-  
+
   UINT8   *KPATIConnectorsData;
 #if defined(MDE_CPU_IA32)
   UINT32  align2;
 #endif
-  
+
   UINTN   KPATIConnectorsDataLen;
 #if defined(MDE_CPU_IA32)
   UINT32  align3;
@@ -585,7 +586,7 @@
 #if defined(MDE_CPU_IA32)
   UINT32  align4;
 #endif
-  
+
   INT32   NrKexts;
   UINT32  align40;
   KEXT_PATCH *KextPatches;   //zzzz
@@ -592,7 +593,7 @@
 #if defined(MDE_CPU_IA32)
   UINT32  align5;
 #endif
-  
+
   INT32    NrForceKexts;
   UINT32  align50;
   CHAR16 **ForceKexts;
@@ -603,7 +604,7 @@
   KERNEL_PATCH *KernelPatches;
   INT32   NrBoots;
   KERNEL_PATCH *BootPatches;
-  
+
 } KERNEL_AND_KEXT_PATCHES;
 
 typedef struct {
@@ -997,5 +998,5 @@
 
 #endif
 /*
- 
+
  EOF */
